pkg/clusteragent/languagedetection/patcher.go

```diff
@@ -137,105 +149,151 @@ func (lp *languagePatcher) run() {
 			if !ok {
 				return
 			}
-			lp.handleEvent(eventBundle)
+			eventBundle.Acknowledge()
+			lp.handleEvents(eventBundle.Events)
 		}
 	}
 }
 
-func (lp *languagePatcher) generateAnnotationsPatch(currentLangs, newLangs langUtil.ContainersLanguages) map[string]interface{} {
-	currentAnnotations := currentLangs.ToAnnotations()
-	targetAnnotations := newLangs.ToAnnotations()
-
-	annotationsPatch := make(map[string]interface{})
+// handleEvents handles events from workloadmeta
+func (lp *languagePatcher) handleEvents(events []workloadmeta.Event) {
+	lp.logger.Tracef("Processing %d events", len(events))
 
-	// All newly detected languages should be included
-	for key, val := range targetAnnotations {
-		annotationsPatch[key] = val
-	}
-
-	// Languages that are no more detected should be removed
-	for key := range currentAnnotations {
-		_, found := annotationsPatch[key]
+	for _, event := range events {
+		owner, err := lp.extractOwnerFromEvent(event)
 
-		if !found {
-			annotationsPatch[key] = nil
+		if err != nil {
+			lp.logger.Errorf("failed to handle event: %v", err)
 		}
+
+		lp.queue.Add(*owner)
 	}
+}
 
-	return annotationsPatch
+func (lp *languagePatcher) extractOwnerFromEvent(event workloadmeta.Event) (*langUtil.NamespacedOwnerReference, error) {
+	entityKind := event.Entity.GetID().Kind
+
+	switch entityKind {
+	case workloadmeta.KindKubernetesDeployment:
+		// Case of deployment
+		deploymentID := event.Entity.(*workloadmeta.KubernetesDeployment).ID
+
+		// extract deployment name and namespace from entity id
+		deploymentIds := strings.Split(deploymentID, "/")
+		namespace := deploymentIds[0]
+		deploymentName := deploymentIds[1]
+
+		// construct namespaced owner reference
+		owner := langUtil.NewNamespacedOwnerReference(
```

Comment:
Perhaps it could be more efficient if `NewNamespacedOwnerReference` would directly return a pointer. What do you think ?

Reply:
Can you please explain how this can be more efficient?

Reply:
Discussed it in private